### ISSUE TITLE
Wrong type of TGS nonce?

### ISSUE THREAD

**Issue by jnshao** [heimdal/heimdal#1219]
The environment is using samba [lorikeet-heimdal](https://gitlab.com/samba-team/devel/lorikeet-heimdal) which is forked from here and is patched for KDC. But the problem was found as kerberos client with a Windows AD. I thouth the bug might exist in both repo due to the same encode implmentation of kerberos TGS nonce.

## Symptom
While using Kerberos authentication, the TGS process may fail on the client side due to a mismatched nonce.

## Reproduce*
It is hard to reproduce because it relies on the random nonce value generated by the `krb5_generate_random_block` function within `get_cred_kdc`. However, it can still be reproduced by directly assigning the nonce value using gdb.

## Bug
The nonce will be encoded in ASN1. And I found that if the nonce is greater than 4286578687 ( > 0xFF7FFFFF), the MSB 8 bits will be discarded.

## Details
Based on the [RFC4120](https://datatracker.ietf.org/doc/html/rfc4120#:~:text=req%2Dbody%20%20%20%20%20%20%20%20%5B4%5D%20KDC%2DREQ%2DBODY%0A%7D-,KDC%2DREQ%2DBODY,-%3A%3A%3D%20SEQUENCE%20%7B%0A%20%20%20%20%20%20%20%20kdc%2Doptions%20%20%20%20%20%20%20%20%20%20%20%20%20%5B0%5D%20KDCOptions%2C%0A%20%20%20%20%20%20%20%20cname), the `nonce` should be `unsigned int`, but the `KDC-REQ-BODY` in [krb5.asn1](https://github.com/heimdal/heimdal/blob/master/lib/asn1/krb5.asn1#L574) is still `Krb5Int32` which is differ to `Krb5UInt32` for ASN1 encode.
```
client side generated the nonce: 4286578688 (0xFF800000)
nonce after encoded in TGS-REQ will be: 8388608 (0x800000)
```

If the nonce type is `Krb5Int32`, it will call `der_put_integer` to encode the value. And if the nonce type is `Krb5UInt32`, it will call `der_put_unsigned` to encode the value. After I change the nonce type of `KDC-REQ-BODY` to `Krb5UInt32` in `krb5.asn1`, the issue was solved.

I wonder is there a compatibility issue with the kerberos implementation? Because I found that [RFC1510](https://datatracker.ietf.org/doc/html/rfc1510#:~:text=encoded%20AP%2DREQ%0A%7D-,KDC%2DREQ%2DBODY,-%3A%3A%3D%20%20%20SEQUENCE%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20kdc%2Doptions) said the nonce should be `INTEGER`. But RFC4120 which I metioned said that it should be `UInt32`.

**Comment by nicowilliams** [comment]
You're correct, this is a bug in `lib/asn1/krb5.asn1`.  Thank you for this excellent report!

**Comment by jnshao** [comment]
@nicowilliams  Thansk for the prompt confirmation!

**Comment by nicowilliams** [comment]
This should get fixed for 8.0.
